From 51aa82678817cd63b8974f945ecbed2e6712a764 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Tue, 16 Jul 2019 17:35:15 -0400
Subject: [PATCH] Make keymaster compatible with latest BoringSSL.

BoringSSL sometimes needs to make FOO_free functions return have int
return values for OpenSSL compatibility. In particular,
https://boringssl-review.googlesource.com/c/boringssl/+/35644 tweaked
ENGINE_free.

include/keymaster/km_openssl/openssl_utils.h currently requires the
return value to be void. Make the template a bit more flexible to
tolerate return values.

This is needed to unblock updates of external/boringssl.

Test: mmma system/keymaster
Change-Id: I4f3110a60ce772fc4ec4ff12f3b1668bb26ff549
---
 include/keymaster/km_openssl/openssl_utils.h | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/include/keymaster/km_openssl/openssl_utils.h b/include/keymaster/km_openssl/openssl_utils.h
index 8586b35..4483045 100644
--- a/include/keymaster/km_openssl/openssl_utils.h
+++ b/include/keymaster/km_openssl/openssl_utils.h
@@ -42,12 +42,13 @@ class EvpMdCtxCleaner {
     EVP_MD_CTX* ctx_;
 };
 
-template <typename T, void (*FreeFunc)(T*)> struct OpenSslObjectDeleter {
+template <typename T, typename FreeFuncRet, FreeFuncRet (*FreeFunc)(T*)>
+struct OpenSslObjectDeleter {
     void operator()(T* p) { FreeFunc(p); }
 };
 
 #define DEFINE_OPENSSL_OBJECT_POINTER(name)                                                        \
-    typedef OpenSslObjectDeleter<name, name##_free> name##_Delete;                                 \
+    typedef OpenSslObjectDeleter<name, decltype(name##_free(nullptr)), name##_free> name##_Delete; \
     typedef UniquePtr<name, name##_Delete> name##_Ptr;
 
 DEFINE_OPENSSL_OBJECT_POINTER(ASN1_BIT_STRING)
@@ -67,7 +68,7 @@ DEFINE_OPENSSL_OBJECT_POINTER(X509)
 DEFINE_OPENSSL_OBJECT_POINTER(X509_EXTENSION)
 DEFINE_OPENSSL_OBJECT_POINTER(X509_NAME)
 
-typedef OpenSslObjectDeleter<BIGNUM, BN_free> BIGNUM_Delete;
+typedef OpenSslObjectDeleter<BIGNUM, void, BN_free> BIGNUM_Delete;
 typedef UniquePtr<BIGNUM, BIGNUM_Delete> BIGNUM_Ptr;
 
 keymaster_error_t ec_get_group_size(const EC_GROUP* group, size_t* key_size_bits);
-- 
2.17.1

